### PR TITLE
removed unecessary fragment

### LINKED ISSUE #4681: Usage of React.Fragment in Container causes Inferno incompatability
# Bug report

First, really thanks a lot for this awesome piece of software 😃 

## Describe the bug

The [`Container`'s `render`](https://github.com/zeit/next.js/blob/canary/lib/app.js#L75-L78) method does not need to be wrapped by [`React.Fragment`](https://reactjs.org/docs/fragments.html#short-syntax).

Just returning `children` is enough, if only one children is ever used.
Are multiple `children` ever considered to be used?

```js
render () {
    const {children} = this.props
    return <>{children}</>
}
```

Since [Inferno does **not** support Fragments](https://github.com/infernojs/inferno/issues/501) yet, this also breaks the inferno example.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:
1. Go to [using-inferno](https://github.com/zeit/next.js/tree/canary/examples/using-inferno)
2. make sure that all missing modules are installed - `inferno-clone-vnode`, `inferno-create-class` and `inferno-create-element`
3. run it.

```sh
{ Error: Inferno Error: createElement() name parameter cannot be undefined, null, false or true, It must be a string, class or function.
    at Object.createElement (/Users/andy/scale-unlimited/webapp_v2/node_modules/inferno-create-element/dist/index.cjs.js:53:15)
    at Container.render (/Users/andy/scale-unlimited/webapp_v2/node_modules/next/dist/lib/app.js:188:29)
    at renderVNodeToString (/Users/andy/scale-unlimited/webapp_v2/node_modules/inferno-server/dist/index.cjs.js:184:41)
    at renderVNodeToString (/Users/andy/scale-unlimited/webapp_v2/node_modules/inferno-server/dist/index.cjs.js:195:20)
    at renderToString (/Users/andy/scale-unlimited/webapp_v2/node_modules/inferno-server/dist/index.cjs.js:308:12)
    at renderPage (/Users/andy/scale-unlimited/webapp_v2/node_modules/next/dist/server/render.js:277:26)
    at Function.getInitialProps (/Users/andy/scale-unlimited/webapp_v2/node_modules/next/dist/server/document.js:67:25)
    at _callee$ (/Users/andy/scale-unlimited/webapp_v2/node_modules/next/dist/lib/utils.js:111:30)
    at tryCatch (/Users/andy/scale-unlimited/webapp_v2/node_modules/regenerator-runtime/runtime.js:62:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/Users/andy/scale-unlimited/webapp_v2/node_modules/regenerator-runtime/runtime.js:296:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/Users/andy/scale-unlimited/webapp_v2/node_modules/regenerator-runtime/runtime.js:114:21)
    at step (/Users/andy/scale-unlimited/webapp_v2/node_modules/@babel/runtime/helpers/asyncToGenerator.js:12:30)
    at _next (/Users/andy/scale-unlimited/webapp_v2/node_modules/@babel/runtime/helpers/asyncToGenerator.js:27:9)
    at /Users/andy/scale-unlimited/webapp_v2/node_modules/@babel/runtime/helpers/asyncToGenerator.js:34:7
    at Promise.F (/Users/andy/scale-unlimited/webapp_v2/node_modules/core-js/library/modules/_export.js:36:28)
    at /Users/andy/scale-unlimited/webapp_v2/node_modules/@babel/runtime/helpers/asyncToGenerator.js:7:12 sourceMapsApplied: true }
```

## Expected behavior

Inferno should work and unnecessary Fragments should not be used.

## System information
 - OS: [e.g. macOS ]
 - Browser latest chrome
 - Version of Next.js: `6.0.3`

## Additional context

Related issue: https://github.com/zeit/next.js/issues/4031

### PR DESCRIPTION
**Pull request by AndyOGo**
fixes #4681 
improves #4031

### CODE REVIEW

**Comment by timneutkens** [comment]
As this will most likely cause `key` errors I guess we should do something like:

`const Fragment = React.Fragment || ({children}) => <div>{children}</div>`

🤔

**Comment by AndyOGo** [comment]
@timneutkens 
Hmh, good point.
At least since React `16` components can render Arrays too, docs:
https://reactjs.org/docs/react-component.html#render

And in my local tests it works with Inferno `5`.

Though I agree it should be tested with other React-like libs like Preact and so on too.
I could also imagine that returning a keyed Array of JSX could be better solution for cross React-like library support, similiar to:

```js
Class Container extends Componenet {
  render () {
     const {children} = this.props

     if (Array.isArray(children) {
       return children.map((child, index) => child.key = index && child)
     }

     return children
   }
 }
```

**Comment by AndyOGo** [comment]
Though I ask my self, will your render multiple children?

**Comment by timneutkens** [comment]
Yeah, my main thinking was libraries that don't support React.Fragment + array return. Though you could easily polyfill `React.Fragment` in that case by doing something like `React.Fragment = ({children}) => children`

File: lib/app.js [hunk 1]
```diff
@@ -74,7 +74,20 @@ export class Container extends Component {
 
   render () {
     const {children} = this.props
-    return <>{children}</>
+
+    // make sure to support all React-like libraries, whether they support Arrays, React.Fragment or not
+    if (Array.isArray(children)) {
+      return children.map((child, index) => {
+        // only set key if it's null or undefined
+        if (child.key == null) {
```
Comment (AndyOGo) [inline lib/app.js]: @timneutkens 
We could also just use a false check, though values like `0` would match that too incorrectly.

File: lib/app.js [hunk 2]
```diff
@@ -74,7 +74,20 @@ export class Container extends Component {
 
   render () {
     const {children} = this.props
-    return <>{children}</>
+
+    // make sure to support all React-like libraries, whether they support Arrays, React.Fragment or not
+    if (Array.isArray(children)) {
```
Comment (AndyOGo) [inline lib/app.js]: @timneutkens 
That does not work for multiple children.
Inferno through:
```sh
rootNode is null
getReactStack@http://localhost:3000/_next/static/commons/main.js:14619:17
reconcileHotReplacement@http://localhost:3000/_next/static/commons/main.js:14948:48
renderReconciler@http://localhost:3000/_next/static/commons/main.js:14962:7
asyncReconciledRender@http://localhost:3000/_next/static/commons/main.js:14970:3
proxiedRender@http://localhost:3000/_next/static/commons/main.js:14259:5
updateClassComponent@http://localhost:3000/_next/static/commons/main.js:7999:24
patchComponent@http://localhost:3000/_next/static/commons/main.js:8043:13
patch@http://localhost:3000/_next/static/commons/main.js:7808:13
render@http://localhost:3000/_next/static/commons/main.js:8402:13
renderReactElement@http://localhost:3000/_next/static/commons/main.js:9245:5
_callee4$@http://localhost:3000/_next/static/commons/main.js:9212:15
tryCatch@http://localhost:3000/_next/static/commons/main.js:16237:37
invoke@http://localhost:3000/_next/static/commons/main.js:16471:22
defineIteratorMethods/</prototype[method]@http://localhost:3000/_next/static/commons/main.js:16289:16
step@http://localhost:3000/_next/static/commons/main.js:166:22
_next@http://localhost:3000/_next/static/commons/main.js:181:9
_asyncToGenerator/</<@http://localhost:3000/_next/static/commons/main.js:188:7
Promise@http://localhost:3000/_next/static/commons/main.js:3255:7
_asyncToGenerator/<@http://localhost:3000/_next/static/commons/main.js:161:12
doRender@http://localhost:3000/_next/static/commons/main.js:9154:10
_callee2$@http://localhost:3000/_next/static/commons/main.js:9060:20
tryCatch@http://localhost:3000/_next/static/commons/main.js:16237:37
invoke@http://localhost:3000/_next/static/commons/main.js:16471:22
defineIteratorMethods/</prototype[method]@http://localhost:3000/_next/static/commons/main.js:16289:16
step@http://localhost:3000/_next/static/commons/main.js:166:22
_next@http://localhost:3000/_next/static/commons/main.js:181:9
_asyncToGenerator/</<@http://localhost:3000/_next/static/commons/main.js:188:7
Promise@http://localhost:3000/_next/static/commons/main.js:3255:7
_asyncToGenerator/<@http://localhost:3000/_next/static/commons/main.js:161:12
render@http://localhost:3000/_next/static/commons/main.js:9032:10
_callee$/<@http://localhost:3000/_next/static/commons/main.js:9003:13
notify/<@http://localhost:3000/_next/static/commons/main.js:11509:16
./node_modules/core-js/library/modules/_collection.js/module.exports/</<@http://localhost:3000/_next/static/commons/main.js:1316:22
notify@http://localhost:3000/_next/static/commons/main.js:11508:7
_callee2$@http://localhost:3000/_next/static/commons/main.js:10963:17
tryCatch@http://localhost:3000/_next/static/commons/main.js:16237:37
invoke@http://localhost:3000/_next/static/commons/main.js:16471:22
defineIteratorMethods/</prototype[method]@http://localhost:3000/_next/static/commons/main.js:16289:16
step@http://localhost:3000/_next/static/commons/main.js:166:22
_next@http://localhost:3000/_next/static/commons/main.js:181:9
run@http://localhost:3000/_next/static/commons/main.js:3153:22
notify/<@http://localhost:3000/_next/static/commons/main.js:3170:30
flush@http://localhost:3000/_next/static/commons/main.js:1960:9
```

File: lib/app.js [hunk 3]
```diff
@@ -74,7 +74,20 @@ export class Container extends Component {
...
+      return <div>{children.map((child, index) => {
```
Comment (AndyOGo) [inline lib/app.js]: @timneutkens 
Fixed above error by wrapping it within a `<div>` as root node

**Comment by AndyOGo** [comment]
@timneutkens 
I have at least working code now, but it could more readable.

But I would like to assure that my understanding is correct.
The `Container` is supposed to be able to render single and multiple `children` right?

Regarding your suggestion I agree that patching `React.Fragment` if it's not defined is the cleanest approach. So I would say we have two cases to patch:

- [ ] Single `chlildren` - this is easy, since it can just be returned, without any wrapping root node, like a `<div>`
- [ ] Multiple `children` - this rather complex, caused it needs proper `key`s at each child and needs a wrapping root node

**Comment by AndyOGo** [comment]
There seems also a `React.Fragment` polyfill package:
https://www.npmjs.com/package/react-dot-fragment

**Comment by HaNdTriX** [comment]
This issue has been resolved in: https://github.com/zeit/next.js/pull/4766

**Pull request closed without merging**